### ISSUE TITLE
Version 1.0

### ISSUE THREAD

**Issue by killercup** [killercup/assert_cli#41]
@epage asked in <https://github.com/killercup/assert_cli/pull/40#issuecomment-331887232>:

> I'm curious, what do you see needing done before moving this to 1.0? While I should get back to my regularly schedule project, I would like to also help work towards more crates hitting 1.0.

@epage good question! I'd like to make the arg parsing a bit leaner (cf. #25 and #32) and maybe improve the error output (#17, #18, #30).

My general goal is to make a library that makes writing tests like [these](https://github.com/killercup/waltz/tree/ef79bf0fc3cfc8dd225cce864e21a91192121afe/waltz_cli/tests) easy and fun. It'll probably involve generating some test boilerplate code, and I'm not entirely sure what I want that API to look like. I'm not even sure if that's in scope for assert_cli. What do you think?

**Comment by epage** [comment]
tl;dr After considering this for the last couple days (and digging into the issues), I feel like PR #43 is sufficient for releasing 1.0.

I feel the main categories of 1.0 blockers
- API changes
  - ~~#32~~ `duct` mostly buys us
    - piping, see my comments on waltz.  
    - advanced command lines parsing, see comment on #25
  - ~~#25~~, see comment on it
  - ~~#18~~ I feel like this could be done without breaking compatibility
- Usable.
  - ~~#17~~ from my own testing, this seems like there is enough output
  - ~~#18~~ I feel like a default output policy should be sufficient for release
  - ~~#30~~ Closed, I don't quite remember enough about it to be actionable.
  - I audited the output in PR #43

RE Waltz's tests

While I see a lot of value in what you're doing with waltz, I feel like that has a longer path to stabilization than the core of assert_cli.

assert_cli does a great job of doing one thing very well, testing short lived programs.  If we take the approach of trying to have assert_cli cover additional use cases, whether its the side effect testing done in waltz or the orchestrating multiple programs with duct's piping, I feel like it might detract from its primary strengths.

Also, thinking about it some more, side effect testing might have value outside of running programs.  Waltz's test harness could be really nice for also wrapping functions.

**Comment by killercup** [comment]
Thanks for the comprehensive list!

Okay, let's keep this crate small! I'll just have to come up with a name
for the crate that does the waltz cli spec thing ;)

I'd like to ping some more of assert_cli's users about feedback on 1.0.

Ed Page <notifications@github.com> schrieb am Mi. 27. Sep. 2017 um 05:01:

**Comment by killercup** [comment]
Pinging some people who used assert_cli in their projects (found using Github search):

We are thinking about doing a 1.0 release and would love to hear your opinion on the crate and ideas for its future!

- @livioribeiro who used it for cargo-readme integration tests
- @bjgill who updated cargo-edit to assert_cli 0.4
- @mattgathu who quite recently used assert_cli in rget
- @jxson who uses it in some tests of to-directory

**Comment by mattgathu** [comment]
Hello @killercup @epage thank you for building such an amazing crate!! It has been really helpful.

I have been using it to test  a rust cli app I'm writing using the [assert_cli::Assert::main_binary](https://docs.rs/assert_cli/0.5.2/assert_cli/struct.Assert.html#method.main_binary) interface. This is part of my contribution for [Increasing Rust's Reach](https://blog.rust-lang.org/2017/06/27/Increasing-Rusts-Reach.html)

I was surprised to find [assert_cli::Assert::main_binary](https://docs.rs/assert_cli/0.5.2/assert_cli/struct.Assert.html#method.main_binary) missing in your docs examples considering how helpful it is!! I had to dig a little deeper into your docs to get it.

I'm planning to: 
* Write a Pull Request to enhance the docs and add example for [assert_cli::Assert::main_binary](https://docs.rs/assert_cli/0.5.2/assert_cli/struct.Assert.html#method.main_binary) and other useful interfaces.
* Write a **How to test Rust CLI apps** blog post, with introduction to writing tests with `assert_cli`

Let me know what you think!

**Comment by mattgathu** [comment]
About future ideas, I have been wondering how feasible it is to test for colored output? 

If my CLI binary prints out colored text, can I test this?

What are your thoughts on this.

**Comment by epage** [comment]
> I was surprised to find assert_cli::Assert::main_binary missing in your docs examples considering how helpful it is!! I had to dig a little deeper into your docs to get it.

Yeah, I suspect thats from it being developed later and that it makes it harder to run the examples as tests.  I agree that more attention should be drawn to it.

> Let me know what you think!

Sounds great!  The main thing to watch out for is not reducing coverage.

**Comment by epage** [comment]
> About future ideas, I have been wondering how feasible it is to test for colored output? If my CLI binary prints out colored text, can I test this?

Most programs detect if they are outputting to a console and don't print color if they are.  Some have a "force color" option.

So either we'd need to figure out how to fake being a console or clients need to have a force color option and set it.

When forced, I assume color just shows up in `stdout` as weird characters.  Can you just manually enter those characters in to test it?  Do you have a proposal for what an API might look like for more advanced color testing?

**Comment by epage** [comment]
Once #46 goes in, should we also need to wait for [Environment](https://github.com/Freyskeyd/environment) to hit 1.0?

**Comment by killercup** [comment]
Good question. I'm okay with publishing this with a re-export of the 0.1. I also think that Environment could be 1.0 already.

**Comment by Freyskeyd** [comment]
I would like to have a way to remove variable from Environment before 1.0.
something like:

```rust
let env: Environment = vec![("foo", "bar"), ("bar", "baz")].into();
...
env.remove("foo");
```

I would also like to have a has/check method:

```rust
let env = Environment::inherit();

if env.has("foo") && env.check_if("bar", "baz") {
    ...
}
```

Having a macro style would be a good way to use it:

```rust
let e = inherit_env!(
    "foo" -> "bar", // <--- don't know if I can do that
    "bar" -> "baz",
);

let e = empty_env!();

assert_cli::Assert::command(&["printenv"])
     .with_env(inherit_env!( "bar" -> "baz" ))
     .stdout().contain("bar=baz")
     .unwrap();

```

**Comment by epage** [comment]
```rust
let e = inherit_env!(
    "foo" -> "bar", // <--- don't know if I can do that
    "bar" -> "baz",
);
```

I think I generally see people use the `=>` syntax when they experiment with these type of things, for example https://github.com/kmcallister/literator

**Comment by Freyskeyd** [comment]
@epage yep that's looking better

**Comment by epage** [comment]
@killercup just recently came across https://github.com/emk/subtitles-rs.  Looks like some of its building blocks are similar to the type of thing you are trying to do with waltz.

I've not explored it enough to know all of the implications of `TestDir`, if that concept is even compatible with my projects.  I could see the idea of splitting the file system tests out into is own crate so the two can be composed together.  I could see using the file system stuff in my lower level integration tests as well.

**Comment by epage** [comment]
I did a usability study of assert_cli at RustBeltRust
- Discoverabiliy:  Searching for "test command line" and "assert command line" did not return results on crates.io but "test cli" and "assert cli" did
- stdout testing: The API looks weird when run through rustfmt because it puts each function on a separate line, so it clouds up what the predicate is operating on
  - Keep it, tools are meant to work for us
  - Go back to having a function per variant
  - Follow `Environment` and take in a predicate parameter
    - Easier to do multiple predicates (option to take in an array)
    - While it helps in places, I found it annoying in clap to access a bunch of different types
    - More linking is needed to help people find all of the relevant predicates

**Comment by killercup** [comment]
> I did a usability study of assert_cli at RustBeltRust

Cool! Tell all the awesome folks in Columbus who may know me I said hi!

> Discoverabiliy

I think we still have a few keywords left we can add to Cargo.toml and "command line" might be a good choice :)

> The API looks weird when run through rustfmt

Ha! I didn't even think of that! But it makes sense, as rustfmt can't know the two calls combined have some special meaning.

> Go back to having a function per variant

I'm totally fine with that actually. We have 4 methods and two implementors, so in the end it's just 8 methods to cover all output asserts.

> Follow Environment and take in a predicate parameter

You mean like `.stderr(Output::contains("foo").and_contains("bar").and_is("baz"))`?

But! At that point me might as well do `.stderr(|x: OutputAssertThingy| -> bool { x.is("bar") || (x.contains("baz") || !x.is("bazinga")) })` to not reinvent boolean expressions. Which I think was something someone was asking for in another issue anyway.

(I'm a bit out of the loop and have only consumed issues on this repo through email for far too long, sorry)

**Comment by epage** [comment]
One thing in favor of accepting predicates as parameters is that someone wanted to pass multiple predicates at once which is more of an issue with `contains` and negated `contains`/`is` and not `is`.

**Comment by epage** [comment]
> But! At that point me might as well do `.stderr(|x: OutputAssertThingy| -> bool { x.is("bar") || (x.contains("baz") || !x.is("bazinga")) })` to not reinvent boolean expressions. Which I think was something someone was asking for in another issue anyway (cf. #57).

I think there is still value in specialized predicates for better error reporting.  We could just use `From` to auto-convert a `Fn` into one of the predicates.

**Comment by epage** [comment]
We've started the major shift to assert_cli 1.0.  assert_cmd is the spiritual successor to what assert_cli today.  We plan to turn assert_cli into a higher level CLI testing crate like [waltz tests](https://github.com/killercup/waltz/tree/ef79bf0fc3cfc8dd225cce864e21a91192121afe/waltz_cli/tests).

The building blocks for assert_cli 1.0 are
- https://github.com/assert-rs/assert_fs/
- https://github.com/assert-rs/assert_cmd/
- https://github.com/assert-rs/predicates-rs/
- https://github.com/Freyskeyd/environment ?

The bad: it will probably take longer to stablize assert_cli because we need to re-evaluate the design and play with ideas

The good: the smaller pieces should be a lot easier to stablize.

**Comment by sunjay** [comment]
If I want to start testing a new CLI application, should I still use this crate? Is there a specific version I should start with?
